crates/containerd-shim-wasmtime/src/executor.rs

```diff
@@ -25,47 +27,42 @@ impl WasmtimeExecutor {
 
 impl Executor for WasmtimeExecutor {
     fn exec(&self, spec: &Spec) -> Result<(), ExecutorError> {
-        let args = oci::get_args(spec);
-        if args.is_empty() {
-            return Err(ExecutorError::InvalidArg);
-        }
-
-        let (mut store, f) = self
-            .prepare(spec, args)
-            .map_err(|err| ExecutorError::Other(format!("failed to prepare function: {}", err)))?;
-
-        log::info!("calling start function");
-        match f.call(&mut store, &[], &mut []) {
-            Ok(_) => std::process::exit(0),
-            Err(_) => std::process::exit(137),
-        };
-    }
-
-    fn can_handle(&self, spec: &Spec) -> bool {
-        // check if the entrypoint of the spec is a wasm binary.
-        let (module_name, _method) = oci::get_module(spec);
-        let module_name = match module_name {
-            Some(m) => m,
-            None => {
-                log::info!("Wasmtime cannot handle this workload, no arguments provided");
-                return false;
+        match can_handle(spec) {
+            Ok(()) => {
+                let args = oci::get_args(spec);
+                if args.is_empty() {
+                    return Err(ExecutorError::InvalidArg);
+                }
+
+                let (mut store, f) = self.prepare(spec, args).map_err(|err| {
+                    ExecutorError::Other(format!("failed to prepare function: {}", err))
+                })?;
+
+                log::info!("calling start function");
+                match f.call(&mut store, &[], &mut []) {
+                    Ok(_) => std::process::exit(0),
+                    Err(_) => std::process::exit(137),
+                };
             }
-        };
-        let path = PathBuf::from(module_name);
+            Err(ExecutorValidationError::CantHandle(_)) => {
+                LinuxContainerExecutor::new(self.stdio.clone()).exec(spec)?;
```

Comment:
I realized this changes the behavior of previous runwasi's mechanism. Previously, wasmtime shim first checks if container is a Linux binary, and then check if it's a wasm container. 

This logic seems to first check if container is a wasm binary, and then a Linux container.

Reply:
That's correct.
That's fine in the wasmedge and wasmtime cases where entrypoint should point to wasm files.
For spin/slight/wws the implementation of the executor should be somewhat different.
If you want take a look at the executor in #250, which keeps the old behaviour we has before this PR.

Reply:
The executor validation of #250 is [here](https://github.com/containerd/runwasi/pull/250/files#diff-25efb300a63bf9e9ac46c7a4f5ed76d477d9ab007003c2d0f808523c3f870085R24-R34)

Reply:
I had a discussion with @cpuguy83 before, and we reached a conclusion that there isn't a robust way to check a wasm text file (wasm binary file could be checkd by verifying its magic number).

Reply:
I agree that the current check (i.e., the module has `.wasm` or `.wat` extension is not robust, but I expect it to work in most cases with wasmedge / wasmtime. But I do think we should find a better way.
For wasmedge / wasmtime, once the module is an OCI artifact the discussion will be different.